clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -123,22 +139,41 @@ public void authenticate() throws IOException {
             return;
 
         switch (saslState) {
+            case SEND_MECHANISM:
+                send(new SaslMechanismRequest(mechanism).toByteBuffer());
+                setSaslState(SaslState.RECEIVE_MECHANISM_RESPONSE);
+                break;
+            case RECEIVE_MECHANISM_RESPONSE:
+                byte[] responseBytes = receiveBytes();
+                if (responseBytes == null)
+                    break;
+                else {
+                    SaslMechanismResponse response;
+                    try {
+                        response = new SaslMechanismResponse(ByteBuffer.wrap(responseBytes));
+                    } catch (Exception e) {
+                        LOG.debug("Invalid SASL mechanism response, server may be using GSSAPI");
+                        throw new SaslException("Invalid SASL mechanism response", e);
+                    }
+                    if (response.errorCode() == SaslMechanismResponse.UNSUPPORTED_SASL_MECHANISM)
+                        throw new SaslException(String.format("Client saslMechanism '%s' not enabled in the server, enabled mechanisms are %s", mechanism, response.enabledMechanisms()));
+                    else if (response.errorCode() != 0)
+                        throw new SaslException(String.format("Unknown error code %d, client mechanism is %s, enabled mechanisms are %s", response.errorCode(), mechanism, response.enabledMechanisms()));
+                    else
+                        setSaslState(SaslState.INITIAL);
+                        // Fall through and start SASL authentication using the configured client mechanism
+                }
             case INITIAL:
-                sendSaslToken(new byte[0]);
-                saslState = SaslState.INTERMEDIATE;
+                sendSaslToken(new byte[0], true);
```

Comment:
It seems that we have an existing issue in the handling of case INITIAL. If we can't completely write all bytes of the sasl token, we have to rely on the next call of authenticate() to finish writing the remaining bytes. However, when the write completes, we will go to the initial state and try to send the token again. It seems that we should be transitioning to the INTERMEDIATE state after the write completes. The same issue seems to exist when transitioning from SEND_MECHANISM to RECEIVE_MECHANISM_RESPONSE, if we can't write all bytes in SaslMechanismRequest in one send call.